src/compile.rs

```diff
@@ -320,6 +324,205 @@ pub fn validate_functions<'a, 'b>(
+            validated_modules =
+                create_validated_modules(m.1, validated_modules.clone(), &modules_vec, types);
+        }
+    }
+    errors.into_result(())
+}
+
+fn find_module_cycles<'a>(
+    module_to_check: &Module,
+    modules_vec: &Vec<(&SimpleFile<String, String>, &'a Module)>,
+    visited_modules: HashSet<&str>,
+) -> Result<(), CascadeErrors> {
+    let mut ret = CascadeErrors::new();
+    for m in &module_to_check.modules {
+        match modules_vec
+            .iter()
+            .find(|&module_info| module_info.1.name == m.as_ref())
+        {
+            Some(module_info) => {
+                if visited_modules.contains(m.as_ref()) || *m == module_to_check.name {
+                    // Cycle
+                    return Err(CascadeErrors::from(
+                        ErrorItem::make_compile_or_internal_error(
+                            "Cycle detected",
+                            Some(module_info.0),
+                            m.get_range(),
+                            "This module contains itself or has a descendent that contains it",
+                        ),
+                    ));
+                }
+                let inner_module = module_info.1;
+
+                let mut new_visited_modules = visited_modules.clone();
+                new_visited_modules.insert(module_to_check.name.as_ref());
+
+                match find_module_cycles(inner_module, modules_vec, new_visited_modules) {
+                    Ok(()) => (),
+                    Err(e) => ret.append(e),
+                }
+            }
+            None => (),
+        }
+    }
+    ret.into_result(())
+}
+
+fn create_validated_modules<'a>(
+    curr_module: &'a Module,
+    mut created_modules: Vec<ValidatedModule>,
+    modules_vec: &Vec<(&SimpleFile<String, String>, &'a Module)>,
+    types: &'a TypeMap,
+) -> Vec<ValidatedModule> {
+    let mut other_modules: Vec<ValidatedModule> = Vec::new();
+    let mut type_infos: Vec<TypeInfo> = Vec::new();
+    if curr_module.modules.len() != 0 {
+        for m in &curr_module.modules {
+            if !&created_modules.iter().any(|x| x.name == m.as_ref()) {
+                match modules_vec.iter().find(|&x| x.1.name == m.as_ref()) {
+                    Some(x) => {
+                        created_modules = create_validated_modules(
+                            x.1,
+                            created_modules.clone(),
+                            modules_vec,
+                            types,
+                        );
+                    }
+                    None => (),
+                }
+            }
+            match &created_modules.iter().find(|x| x.name == m.as_ref()) {
+                Some(x) => other_modules.push((**x).clone()),
+                None => (),
+            }
+        }
+    }
+    let name = &curr_module.name;
+    for d in &curr_module.domains {
+        match types.get(d.as_ref()) {
```

Comment:
I think it's a little strange to call types.get() in the validation, and then again when creating the ValidatedModule.  Can we not save the TypeInfos from when we validate?